### PR TITLE
[CSL-2089] Fix diffusion layer

### PR DESCRIPTION
**Pull request by volhovm**
This PR attempts to cleanup some things that became redundant after diffusion layer was merged, also it tries to fix several important bugs that were not noticed at the review stage.

At the time of writing, I've pushed two commits. One removes copy-pasted logic that was moved to the diffusion layer, but wasn't cleaned up from `Block.Networking`. Second fixes exception handling that violate the exceptions guideline (`docs/exceptions.md`). It also renames some things and removes `getChainFrom` (which must be implemented in a completely different way and wasn't used anywhere). 

As for bugs, there are two of them:
1. `getHashesRange` is used without `depthLimit` in diffusion layer. This means everybody can query the whole blockchain at once, which is an easy way to attack the node. This changed wasn't discuss explicitly, so it's a bug unless it's decided to think otherwise.
2. More serious one: block/headers communication is broken in case of rollbacks. It was decided to move all the checks out of the diffusion layer, but one important bit was missed: `classifyHeaders` was returning LCA header which then was helping to form `MsgGetBlocks`. Now, we request the very same headers server returned to us after `MsgGetHeaders` call, which is a mistake, because checkpoints are distributed exponentially over last `k` blocks (so some blocks are not included). Example: client has `A B C D E F G H` (oldest first), server has `A B C D E' F' G' H'`, client sends checkpoints set `{H G E A}` and thus server wil return `B..H'`, then the corresponding blocks will be fetched. But in fact we only need `E'..H'`. This comparison logic was previously made in `classifyHeaders`. I must also admit that there is an extra check (taking LCA in `handleBlocks`) that prevents client to fail when he tries to apply unexpectedly big list of blocks (with already known prefix). But facts are still left: protocol is used incorrectly, leading to non-optimal data transfer in case of forks.

@avieth Please tell your thoughts on this and let's decide what to do. 

IMHO:
* (1) behavior should be restored, it's easy to do so
* (2) should be left as it is, because I think the protocol format will be changed to a more compact one (one request-response instead of two) anyway. Remember though, that neither cheap to increase the size of checkpoints set on the client side as it leads to slower request processing on the server side (every checkpoint must be checked).

### CODE REVIEW

File: block/src/Pos/Block/Logic/Header.hs [hunk 1]
```diff
@@ -352,29 +357,39 @@ getHeadersOlderExp upto = do
...
     => Maybe Word
     -> HeaderHash
     -> HeaderHash
-    -> m (Either Text (OldestFirst NE HeaderHash))
-getHeadersRange depthLimitM older newer | older == newer = runExceptT $ do
+    -> m (OldestFirst NE HeaderHash)
+getHashesRange depthLimitM older newer | older == newer = do
     unlessM (isJust <$> DB.getHeader newer) $
-        throwError "getHeadersRange: can't find newer-older header"
+        throwGHR "can't find newer-older header"
```
Comment (avieth) [inline block/src/Pos/Block/Logic/Header.hs]: Is this really an error? Isn't it a completely normal thing that can be handled in a non-exceptional way?

File: block/src/Pos/Block/Logic/Header.hs [hunk 2]
```diff
@@ -352,29 +357,39 @@ getHeadersOlderExp upto = do
...
     whenJust depthLimitM $ \depthLimit ->
         when (depthLimit < 1) $
-        throwError $
-        sformat ("getHeadersRange: depthLimit is "%int%
+        throwGHR $
```
Comment (avieth) [inline block/src/Pos/Block/Logic/Header.hs]: Again, really an error? Can't we handle it non-exceptionally?

File: block/src/Pos/Block/Logic/Header.hs [hunk 3]
```diff
@@ -384,13 +399,13 @@ getHeadersRange depthLimitM older newer = runExceptT $ do
...
-        throwError $
-        sformat ("getHeadersRange: newer and older headers have "%
+        throwGHR $
+        sformat ("newer and older headers have "%
                  "the same difficulty, but are not equal. "%newerOlderF)
                 newerHd olderHd
     when (newerD < olderD) $
-        throwError $
-        sformat ("getHeadersRange: newer header is less dificult than older one. "%
+        throwGHR $
```
Comment (avieth) [inline block/src/Pos/Block/Logic/Header.hs]: This and the `throwGHR` on line 402 seem misplaced. Why are doing these checks here? Surely this function is taking on too much responsibility. Why can't it just fetch the headers and leave it at that?

File: lib/src/Pos/Logic/Full.hs [hunk 4]
```diff
@@ -123,23 +117,19 @@ logicLayerFull jsonLogTx k = do
         getBlockHeader :: HeaderHash -> m (Maybe BlockHeader)
         getBlockHeader = DB.getHeader
 
+        getHashesRange
+            :: HeaderHash
+            -> HeaderHash
+            -> m (OldestFirst NE HeaderHash)
+        -- TODO CSL-2089 bug: should be Just k, but not
+        -- Nothing. Probably, listener logic is broken.
+        getHashesRange = Block.getHashesRange Nothing
```
Comment (avieth) [inline lib/src/Pos/Logic/Full.hs]: Should it be `Just k`? We plan to improve the horribly slow block sync by allowing a client to fetch *more* than `k` blocks (i.e. the whole chain) at once.

File: lib/src/Pos/Logic/Types.hs [hunk 5]
```diff
@@ -27,47 +26,33 @@ import           Pos.Util.Chrono (NE, NewestFirst, OldestFirst)
...
+    { -- | The stakeholder id of our node.
       ourStakeholderId   :: StakeholderId
-      -- Get a block, perhaps from a database.
+      -- | Get a block, perhaps from a database.
     , getBlock           :: HeaderHash -> m (Maybe Block)
-      -- Stream blocks from first hash to second hash.
-      -- Conduit is chosen mainly due to precedent: it's already used in
-      -- cardano-sl.
-    , getChainFrom       :: HeaderHash
-                         -> Source m Block
```
Comment (avieth) [inline lib/src/Pos/Logic/Types.hs]: Why did you delete this? Please restore it.

Comment (avieth) [reply lib/src/Pos/Logic/Types.hs]: Oh

> removes getChainFrom (which must be implemented in a completely different way and wasn't used anywhere).

It's going to be used. CSL-2148

Comment (volhovm) [reply lib/src/Pos/Logic/Types.hs]: Yes, my idea is that:
1. All DB traversals should be conduits (it's easy to do it).
2. `getHeadersFromManyTo` and `getHashesRange` should be too, also I think first one should be implemented in terms of the second one. 

So `getChainFrom` is an unused code with implementation that will be totally different, that's why i deleted it. I suggest we keep things simple and move on incrementally. It's 5m to somebody to return it back, but it takes much more time for me to figure out which functionality is actually needed and which is unused.

File: lib/src/Pos/Logic/Full.hs (hunk 4 above)
Comment (volhovm) [reply lib/src/Pos/Logic/Full.hs]: In fact, it's simpler to get rid of this parameter at all. I just wanted to consult whether it is a mistake or not.

**Comment by volhovm** [comment]
@avieth any comments on the (2) bug?

File: block/src/Pos/Block/Logic/Header.hs (hunk 1 above)
Comment (int-index) [reply block/src/Pos/Block/Logic/Header.hs]: The updated code follows the guidelines in https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md unlike the old code.

The reasoning behind the current design choice in the guidelines is that we want only a single way to throw exceptions in any part of code - it could be `ExceptT` in pure code, or `MonadThrow` in (potentially) impure code. This guarantees that we can safely use `bracket` everywhere, and that there's never need to use both `catch` and `catchError`.

This code was using `throwError`, but it wasn't pure, so it was rightly rewritten. If you disagree with the use of `throwM` here, the alternative is to return `Either`, but without `runExceptT` which creates a secondary exit path from any part of the code.

Additionally, we don't use `Text` to represent errors (or "normal things", as you refer to them above) anymore, and the updated code at least uses a newtype (ideally it should be an ADT).

File: block/src/Pos/Block/Logic/Header.hs (hunk 2 above)
Comment (int-index) [reply block/src/Pos/Block/Logic/Header.hs]: Again, `throwError` that was there before is not "non-exceptionally" either, we should either use `throwM` here or a non-monadic way to handle this codepath.

**Comment by avieth** [comment]
> (1) behavior should be restored, it's easy to do so

I'm not so sure. IIRC the `depthLimit` was in practice determined by the recovery headers limit parameter in the configuration. It was in practice 2200. This also was used to compute the size limit of `MsgHeaders`. Its effectiveness at mitigating a denial of service attack is modest: the attacker can just spam `MsgGetHeaders` requests which it knows will lead to 2200 headers in the response. Is that significantly less bad than allowing an attacker to request the whole blockchain?

**Comment by avieth** [comment]
>  any comments on the (2) bug?

I'm trying to understand it. We give the server our checkpoints and tip. Server responds with a list of hashes corresponding to a chain, where one end is in our checkpoint set, and the other is the tip we requested. We used to follow that chain to get the LCA with our chain, potentially coming up with a smaller chain to request.

Another option is to have the server do this, no? Instead of giving back all of the headers in a list it could give two hahses, oldest and newest, where the oldest lies in-between two of our checkpoints, or coincides with one of them. That's to say, the *server* computes the shortest chain that the client needs. Even better, the server could *not* respond with the two hashes, but instead with the actual shorted blockchain that the client needs.

**Comment by volhovm** [comment]
>Is that significantly less bad than allowing an attacker to request the whole blockchain?

We don't have any streaming style, which means that everything will be first load into the RAM. It means that if all the hashes in the chain are big, it's an easy way to fill up all the server RAM. I guess this is much easier attack than just spamming `MsgGetBlocks` (`getHeaderHashes` is used there), which would target lot of IO, which in turn would slow down the whole system. We use blake2b, so `256` bytes per 400k blocks (this is current chain length I believe) is 100Mb, which is huge. 

>where the oldest lies in-between two of our checkpoints, or coincides with one of them

How would you compute this on the server side? It doesn't have enough info about client side chain. 

One more note: previously, sending headers list to client (instead of two headers as you suggest) was used to check that the server side is not trying to fool us, as we can perform pretty good headers validation. It's not that trivial to tamper, because headers have slot leaders which we already know in advance, and they also have signatures.

**Comment by avieth** [comment]
> > where the oldest lies in-between two of our checkpoints, or coincides with one of them
>
> How would you compute this on the server side? It doesn't have enough info about client side chain.

Right indeed, it would have to give one of the actual checkpoints.

How about a newest-to-oldest approach? Server claims to have a more difficult tip than client's, so client asks for the chain *from the tip, backwards*. Server can stream the blocks one-at-a-time with rate controlled by the client, and client can stop as soon as it meets its existing chain.

> One more note: previously, sending headers list to client (instead of two headers as you suggest) was used to check that the server side is not trying to fool us, as we can perform pretty good headers validation. It's not that trivial to tamper, because headers have slot leaders which we already know in advance, and they also have signatures.

Some checks can be done in the streaming style as well (with a bunch of state-related hazards of course). For each new block, assuming we have enough information about its epoch, we can check that it's signed by the right leader, and that the block we received before it has the appropriate hash, etc. and stop the streaming if this is ever violated.

**Comment by volhovm** [comment]
>How about a newest-to-oldest approach? 

Very controversial. 
* We can't immediately data that we get. We also can't verify it. 
* It's O(n) versus O(logn) in existing approach, and I don't see any advantages. I think assuming that server side has returned correct chain (without verifying) and finding LCA on the client in the unsafe way (meaning that we can fail to find it) is faster and better.

>Some checks can be done in the streaming style as well 

This is not _exactly_ true, but can be in theory. And I'd like it to be true.

What about (1)? Can we bring the check back for now and later rework this when switching to the never protocol version?

**Comment by avieth** [comment]
> It's O(n) versus O(logn) in existing approach, and I don't see any advantages. I think assuming that server side has returned correct chain (without verifying) and finding LCA on the client in the unsafe way (meaning that we can fail to find it) is faster and better.

I don't see the linear v. logarithmic distinction. In either case, we have to transmit the same number of blocks.

> What about (1)? Can we bring the check back for now and later rework this when switching to the never protocol version?

Ok I'll look into this. To be sure I understand: after requesting the headers, find the newest hash in the list such that we have that block, and then request from the next header to the tip?

**Comment by volhovm** [comment]
>I don't see the linear v. logarithmic distinction. In either case, we have to transmit the same number of blocks.

Yes, there's obviously the same number of blocks. We try to minimize the data to get `[b1,bn]` which we should then query. Indeed, my mistake, in your top-to-bottom solution the number of headers sent from server to client is O(n) as it is in worst case in the current solution. What we _could_ do is to send client `[c1..c2]`, where `c1` is the newest checkpoint present on the server and `c2` is the oldest checkpoint (after `c1`) which is not present on it. Then client can find LCA among `[c1..c2]`. This is still O(n) data sent, but much less (n/2 in the worst case). 

Top-to-bottom solution seems fine actually, but i suggest to think about everything in a separate ticket, because we better batch all the protocol changes to make things optimal. I would also be happy to see some simplification of this two-request scheme if it's possible. What i suggest now is to take LCA between querying headers and asking for blocks, doing no checks. 

>after requesting the headers, find the newest hash in the list such that we have that block, and then request from the next header to the tip?

Yes, this is a correct understanding, but it's not relevant to (1). (1) is about restricting traversals. See this comment from me again (first part of it): https://github.com/input-output-hk/cardano-sl/pull/2460#issuecomment-363836042

**Comment by avieth** [comment]
> Yes, this is a correct understanding, but it's not relevant to (1). (1) is about restricting traversals. See this comment from me again (first part of it): #2460 (comment)

Oh right, got my numbers mixed up. I was talking about 2.

File: block/src/Pos/Block/Logic/Header.hs (hunk 1 above)
Comment (avieth) [reply block/src/Pos/Block/Logic/Header.hs]: > Additionally, we don't use Text to represent errors (or "normal things", as you refer to them above) anymore, and the updated code at least uses a newtype (ideally it should be an ADT).

Sure I'm all for that. In principle, though, I think many of the `throw*`s seen in this patch aren't really exceptions, they're normal outcomes. This line in particular: a header hash was given, but it's not in the database. Ok, that's an expected outcome, and the caller can deal with it normally. We wouldn't promise the caller a total function from header hash to blocks, so we should't pretend we have by giving such a type.

Comment (int-index) [reply block/src/Pos/Block/Logic/Header.hs]: The normal outcomes that short-circuit the computation (thus can be, theoretically, represented by `MonadThrow` or `MonadError`) go under the name of "regular errors" in the guidelines. You might argue that it's a misnomer, but it doesn't change the essence of the question.

**Comment by avieth** [comment]
Actually I believe issue 1 isn't a problem. `getHeadersFromManyTo` already has the `recoveryHeadersMessage` built-in:

```Haskell
let loadUpCond (headerHash -> curH) h =
        curH /= startHash && h < recoveryHeadersMessage
GS.loadHeadersUpWhile newestCheckpoint loadUpCond
```

@volhovm Should be fine, right?

**Comment by volhovm** [comment]
@avieth Not really. Indeed, when server is asked about headers (given checkpoints), he returns no more than `recoveryHeadersMessage`. So yes, protocol works as expected next -- client queries about the same number of headers. But again, without `getHeaderHashes` limitation anybody can ask server to retrieve `[b0..bs]` , where `b0` is the very first block, and `bs` is some block having difficulty 10^6+, and node will load everything into memory. Hashes are 256 bytes, so 10^6 hashes is about 250 Mb. That's a very easy OOM, takes malicious client just a bunch of simultaneous calls to achieve node crash.

We can later remove this limit when we switch to conduit-based streaming (which will not load all the data into memory at once).

**Comment by avieth** [comment]
@volhovm There's 2 listeners which load up block (header) data: get headers and get blocks.

The former will use `getBlockHeaders` from the logic layer, which has the depth limit built in (as I showed above) so that's fine, no attacker can make the server fill up its heap beyond some bound in this way.

Maybe the get blocks listener is vulnerable? It uses `getBlockHeaders'`, a variant that takes two particular ends point hashes rather than a list of checkpoints. The server will get all hashes between these two points, so yes indeed we need a limit. Full logic layer uses `getHeadersRange Nothing`, so we have to change that. Is this what you meant?

**Comment by avieth** [comment]
https://github.com/input-output-hk/cardano-sl/pull/2484

**Comment by volhovm** [comment]
> Is this what you meant?

Yes, I'll fix it in this PR.

**Comment by volhovm** [comment]
@avieth I've restored the behaviour with LCA on (2). Please review this PR again.

File: block/src/Pos/Block/Logic/Header.hs (hunk 1 above)
Comment (avieth) [reply block/src/Pos/Block/Logic/Header.hs]: It's confusing. Here I have some `m` that could be and probably is `IO`-capable. What does "throwing" mean exactly? It suggests an exception, as in a real true exception (the kind that only makes sense in `IO`) but it could also be a "normal" short-circuit.

Anyway, it's beyond the scope of this patch I guess. I think the true issue is the lack of good separation of pure and impure programs. If we had that, then there would be no ambiguity about what "throwing" means.

File: lib/src/Pos/Diffusion/Full/Block.hs [hunk 6]
```diff
@@ -124,11 +120,27 @@ getBlocks logic enqueue nodeId tipHeader checkpoints = do
...
     -- the block itself.
     blocks <- if singleBlockHeader
-              then requestBlocks (NewestFirst (pure tipHeader))
-              else requestHeaders >>= requestBlocks
+              then requestBlocks (OldestFirst (one tipHeader))
+              else twoPhaseAction
     pure (OldestFirst (reverse (toList blocks)))
   where
 
+    twoPhaseAction = do
```
Comment (avieth) [inline lib/src/Pos/Diffusion/Full/Block.hs]: Needs a comment and a better name if you can think of one. How about `getAndClassifyHeaders`, then take out `requestBlocks` from the end and put it back up on line 124?

File: lib/src/Pos/Diffusion/Full/Block.hs [hunk 7]
```diff
@@ -124,11 +120,27 @@ getBlocks logic enqueue nodeId tipHeader checkpoints = do
...
+        headers <- toOldestFirst <$> requestHeaders
+        getLcaMainChain logic headers >>= \case
```
Comment (avieth) [inline lib/src/Pos/Diffusion/Full/Block.hs]: How much work would it be to make something like `getLcaMainChain` that, instead of giving the lca hash, gives the suffix that we compute after this? So that we wouldn't have to `dropUntilLca` again? This could save a traversal and eliminate that error case on line 140.

File: lib/src/Pos/Logic/Full.hs [hunk 8]
```diff
@@ -123,25 +117,22 @@ logicLayerFull jsonLogTx k = do
         getBlockHeader :: HeaderHash -> m (Maybe BlockHeader)
         getBlockHeader = DB.getHeader
 
+        getHashesRange
+            :: Maybe Word -- ^ Optional limit on how many to pull in.
+            -> HeaderHash
+            -> HeaderHash
+            -> m (OldestFirst NE HeaderHash)
+        getHashesRange = Block.getHashesRange
```
Comment (avieth) [inline lib/src/Pos/Logic/Full.hs]: Good this is a better name.

File: block/src/Pos/Block/Logic/Header.hs (hunk 1 above)
Comment (volhovm) [reply block/src/Pos/Block/Logic/Header.hs]: @avieth this patch tries to fix problems that were missed in PR 2154: https://github.com/input-output-hk/cardano-sl/pull/2154/#discussion_r163384764

* Code is violating exceptions guideline
* It wasn't fixed, so now it is

My patch may be confusing, but it is only because exceptions handling is hard. In this particular place there is no good separation between request exceptions and everything else, pure vs impure, so I just follow the guideline written by @int-index. You're more than welcomed to read and argue with it, but for now -- standards first.

File: lib/src/Pos/Diffusion/Full/Block.hs (hunk 6 above)
Comment (volhovm) [reply lib/src/Pos/Diffusion/Full/Block.hs]: Sure, I will.

File: lib/src/Pos/Diffusion/Full/Block.hs (hunk 7 above)
Comment (volhovm) [reply lib/src/Pos/Diffusion/Full/Block.hs]: `dropUntilLca` usually does exactly 0 steps, because usually headers list is already `[c1..cn]`, where `c1` is a child of our tip `c0`, so it's a direct continuation. I don't see any need to optimize here, it's a _really_ rare case that this `dropUntilLca` will actually do something, because rollbacks are pretty rare. And if they are, it's anyway super quick and happens only once (there can't be more than one serious rollback).

File: block/src/Pos/Block/Logic/Header.hs (hunk 3 above)
Comment (volhovm) [reply block/src/Pos/Block/Logic/Header.hs]: Well, underlying traversal functions require me to put in some sane data, so this is a perfect place to check check its validity. `getHashesRange` is _actually_ a listener, and it's not used anywhere else, so the only difference between doing checks here and doing them outside is how do we handle errors. Here there's `GetHashesRangeException` already that will carry all the info about bad user request. If we do checks outside, we need yet another exception type. Also these checks are simple/fast.

**Comment by volhovm** [comment]
@avieth Check out https://github.com/input-output-hk/cardano-sl/pull/2460/commits/fb50e112c8ed5cd256e936da252737388fa7e682

File: block/src/Pos/Block/Logic/Header.hs (hunk 1 above)
Comment (avieth) [reply block/src/Pos/Block/Logic/Header.hs]: Right I won't let it stand in the way of merging this, it's a separate issue. It's not really about exception handling guidelines, it's about when to use exceptions at all.

File: block/src/Pos/Block/Logic/Header.hs (hunk 3 above)
Comment (avieth) [reply block/src/Pos/Block/Logic/Header.hs]: > getHashesRange is actually a listener, and it's not used anywhere else, so the only difference between doing checks here and doing them outside is how do we handle errors

I disagree. Sure its only user at the moment is a listener, but it has no intrinsic relation to networking / listener / diffusion layer. One could argue it's a database component but even that is a bit dubious. `getHashesRange` is what its type says it is: given two hashes, get (somehow) a list of hashes. Doesn't matter where the inputs came from, doesn't matter where the output goes.

File: explorer/src/Pos/Explorer/Socket/Methods.hs [hunk 9]
```diff
@@ -358,10 +359,10 @@ getBlundsFromTo
     :: forall ctx m . ExplorerMode ctx m
     => HeaderHash -> HeaderHash -> m (Maybe [Blund])
 getBlundsFromTo recentBlock oldBlock =
-    DB.getHeadersRange Nothing oldBlock recentBlock >>= \case
+    tryAny (DB.getHashesRange Nothing oldBlock recentBlock) >>= \case
```
Comment (avieth) [inline explorer/src/Pos/Explorer/Socket/Methods.hs]: Perhaps it would be better to not catch the exception? But right, `getHashesRange` is given a type that doesn't even admit the possibility of a normal failure. I'll write about that in a separate comment.

I take issue with this one because we're conflating two very different reasons for failure:

1. Something normal/expected, like not knowing one of the endpoints.
2. Something *actually* exceptional, like the database being corrupt or a failure to open some file it uses.

File: lib/src/Pos/Logic/Types.hs [hunk 10]
```diff
@@ -27,53 +25,39 @@ import           Pos.Util.Chrono (NE, NewestFirst, OldestFirst)
...
-                         -> HeaderHash
-                         -> HeaderHash
-                         -> m (Either GetBlockHeadersError (OldestFirst NE HeaderHash))
-      -- Get the current tip of chain.
-      -- It's not in Maybe, as getBlock is, because really there should always
-      -- be a tip, whereas trying to get a block that isn't in the database is
-      -- normal.
+                         -> m (NewestFirst NE BlockHeader)
+      -- | Compute LCA with the main chain.
+    , getLcaMainChain    :: OldestFirst NE BlockHeader -> m (Maybe HeaderHash)
```
Comment (avieth) [inline lib/src/Pos/Logic/Types.hs]: Notice that this type expresses the normal failure case: there's no LCA, so you get `Nothing`.

Compare at `getHashesRange`, which completely denies the normal failure case of not knowing one of the endpoints.

File: lib/src/Pos/Logic/Types.hs [hunk 11]
```diff
@@ -27,53 +25,39 @@ import           Pos.Util.Chrono (NE, NewestFirst, OldestFirst)
...
     , getBlockHeader     :: HeaderHash -> m (Maybe BlockHeader)
-      -- Inspired by 'getHeadersFromManyTo'.
-      -- Included here because that function is quite complicated; it's not
-      -- clear whether it can be expressed simply in terms of getBlockHeader.:q
+      -- TODO CSL-2089 use conduits in this and the following methods
+      -- | Retrieve block header hashes from specified interval.
+    , getHashesRange     :: Maybe Word -- ^ Optional limit on how many to bring in.
+                         -> HeaderHash
+                         -> HeaderHash
+                         -> m (OldestFirst NE HeaderHash)
```
Comment (avieth) [inline lib/src/Pos/Logic/Types.hs]: Change it to `Either GetHashesRangeError (OldestFirst NE HeaderHash)` or `Maybe (OldestFirst NE HeaderHash)`.

File: lib/src/Pos/Logic/Types.hs [hunk 12]
```diff
@@ -27,53 +25,39 @@ import           Pos.Util.Chrono (NE, NewestFirst, OldestFirst)
...
-      -- or similar).
-    , getBlockHeaders'   :: Maybe Word -- ^ Optional limit on how many to bring in.
-                         -> HeaderHash
-                         -> HeaderHash
-                         -> m (Either GetBlockHeadersError (OldestFirst NE HeaderHash))
-      -- Get the current tip of chain.
-      -- It's not in Maybe, as getBlock is, because really there should always
-      -- be a tip, whereas trying to get a block that isn't in the database is
-      -- normal.
+                         -> m (NewestFirst NE BlockHeader)
```
Comment (avieth) [inline lib/src/Pos/Logic/Types.hs]: Change it to `m (Either GetBlockHeadersError (NewestFirst NE BlockHeader))`.

File: block/src/Pos/Block/Logic/Header.hs (hunk 1 above)
Comment (avieth) [reply block/src/Pos/Block/Logic/Header.hs]: I take it back, this *is* a problem. Discussion here https://github.com/input-output-hk/cardano-sl/pull/2492

**Comment by volhovm** [comment]
@avieth I've returned back `Either e res`, though `e` is now a datatype. Can you please review this again?

**Comment by avieth** [comment]
> @avieth I've returned back Either e res, though e is now a datatype. Can you please review this again?

Ok looking now.

File: block/src/Pos/Block/Logic/Header.hs [hunk 13]
```diff
@@ -7,9 +7,12 @@ module Pos.Block.Logic.Header
        , classifyNewHeader
        , ClassifyHeadersRes (..)
        , classifyHeaders
+
+       , GetHeadersFromManyToError
```
Comment (avieth) [inline block/src/Pos/Block/Logic/Header.hs]: Should export constructors too, no?

File: block/src/Pos/Block/Logic/Header.hs [hunk 14]
```diff
@@ -7,9 +7,12 @@ module Pos.Block.Logic.Header
...
        , getHeadersFromManyTo
        , getHeadersOlderExp
-       , getHeadersRange
+       , GetHashesRangeError
```
Comment (avieth) [inline block/src/Pos/Block/Logic/Header.hs]: `(..)`

File: block/src/Pos/Block/Logic/Header.hs [hunk 15]
```diff
@@ -244,6 +247,9 @@ classifyHeaders inRecovery headers = do
                           depthDiff blkSecurityParam
             | otherwise -> CHsValid lcaChild
 
+
+data GetHeadersFromManyToError = GHFBadRequest Text deriving (Show,Generic)
```
Comment (avieth) [inline block/src/Pos/Block/Logic/Header.hs]: Minor nit: networking language. `GHFBadInput`, `GHFUnknownHeader` or similar instead? Sure, this is *used* by the diffusion layer / networking but it's intended to be modular.

File: block/src/Pos/Block/Logic/Header.hs [hunk 16]
```diff
@@ -354,29 +361,40 @@ getHeadersOlderExp upto = do
                 | otherwise = selGo es ii $ succ skipped
         in selGo elems ixs 0
 
+
+data GetHashesRangeError = GHRBadRequest Text deriving (Show,Generic)
```
Comment (avieth) [inline block/src/Pos/Block/Logic/Header.hs]: Similar nit to the naming of `GHFBadRequest`.

File: block/src/Pos/GState/BlockExtra.hs [hunk 17]
```diff
@@ -177,26 +175,6 @@ loadBlocksUpWhile
...
-    mBlk <- lift $ getBlock fromH
-    case mBlk of
-        Nothing  -> pure ()
-        Just blk -> yield blk >> continue blk
-  where
-    continue blk = do
-        mNextH <- lift $ resolveForwardLink blk
-        case mNextH of
-            Nothing    -> pure ()
-            Just nextH -> blocksSourceFrom nextH
```
Comment (avieth) [inline block/src/Pos/GState/BlockExtra.hs]: @karknu relevant to what we were talking about earlier.

Comment (avieth) [reply block/src/Pos/GState/BlockExtra.hs]: It's deleted here because it's not used yet, but you will probably want to add something similar. Pipe or conduit or similar, whatever you think is appropriate

File: explorer/src/Pos/Explorer/Socket/Methods.hs [hunk 18]
```diff
@@ -358,10 +359,10 @@ getBlundsFromTo
     :: forall ctx m . ExplorerMode ctx m
     => HeaderHash -> HeaderHash -> m (Maybe [Blund])
 getBlundsFromTo recentBlock oldBlock =
-    DB.getHeadersRange Nothing oldBlock recentBlock >>= \case
+    tDB.getHashesRange Nothing oldBlock recentBlock >>= \case
```
Comment (avieth) [inline explorer/src/Pos/Explorer/Socket/Methods.hs]: Typo? `tDB`

File: lib/src/Pos/Diffusion/Full/Block.hs [hunk 19]
```diff
@@ -124,11 +120,30 @@ getBlocks logic enqueue nodeId tipHeader checkpoints = do
...
+                -- Headers list is (oldest to newest)
+                -- [n1,n2,...nj,lca,nj+2,...nk] we drop [n1..lca] and
+                -- return [nj+2..nk], as we already have lca in our
+                -- local db. Usually this function does 1 iterations
+                -- as it's a common case that [n1..nj] is absent and
+                -- lca is the oldest header.
+                let dropUntilLca = NE.dropWhile (\h -> h ^. prevBlockL /= lca)
+                case nonEmpty (dropUntilLca $ getOldestFirst headers) of
+                    Nothing -> throwM $ DialogUnexpected $
+                                   "All headers are older than LCA, nothing to query"
```
Comment (avieth) [inline lib/src/Pos/Diffusion/Full/Block.hs]: I'm not demanding this change here but I just want to ask, do we agree that this would be better as an explicit `Left (DialogUnexpected ...)`? So that the type of `getBlocks` would end in `d (Either <error type> OldestFirst [] Block)`?

File: block/src/Pos/Block/Logic/Header.hs (hunk 15 above)
Comment (volhovm) [reply block/src/Pos/Block/Logic/Header.hs]: Yes, fixed. It wasn't intentional, `BadInput` is better.

File: lib/src/Pos/Diffusion/Full/Block.hs (hunk 19 above)
Comment (volhovm) [reply lib/src/Pos/Diffusion/Full/Block.hs]: This should be a completely different story, I think. Currently, `retrievalWorker` expects an exception as a sign of "something went wrong", so we eventually throw even if we get `Left` somewhere. So at this point, no. But I'd took a look at exception handling in retrieval worker. 

The suggestion itself is something to think about. At a logic layer, I don't really care whether it was a "normal" exception like "somebody sent us trash" or "exceptional" like "my computer is on fire". But maybe I'm missing something.

File: block/src/Pos/GState/BlockExtra.hs (hunk 17 above)
Comment (volhovm) [reply block/src/Pos/GState/BlockExtra.hs]: Me? Or @karknu ? :)

I'll explain again: i have a vision^{tm} of how conduits should be structured. Namely, all traversals should be conduit. And when they are, the particular deleted function will become redundant. Also it wasn't used and it's always available in git history, just in case.

Comment (avieth) [reply block/src/Pos/GState/BlockExtra.hs]: Nope, I meant @karknu . We were discussing block streaming earlier today.

**Comment by volhovm** [comment]
@avieth https://github.com/input-output-hk/cardano-sl/pull/2460/commits/f4f926bc1d4c6390368fbc4f0df6aa60f826ab74
